middle_end/flambda2/tests/algorithms/patricia_tree_tests.ml

```diff
@@ -0,0 +1,915 @@
+
+  let update_other k1 k2 f m =
+    k1 = k2 || Map.find_opt k1 (Map.update k2 f m) =? Map.find_opt k1 m
+
+  let singleton_valid k v = Map.valid (Map.singleton k v)
+
+  let singleton k1 k2 v =
+    Map.find_opt k1 (Map.singleton k2 v) =? if k1 = k2 then Some v else None
+
+  let remove_valid k m = Map.valid (Map.remove k m)
+
+  let remove_same k m = Map.find_opt k (Map.remove k m) =? None
+
+  let remove_other k1 k2 m =
+    k1 = k2 || Map.find_opt k1 (Map.remove k2 m) =? Map.find_opt k1 m
+
+  let merge_valid f m1 m2 = Map.valid (Map.merge f m1 m2)
+
+  let merge f m1 m2 =
+    let mm = Map.merge f m1 m2 in
+    let correct_for_key k =
+      Map.find_opt k mm =? f k (Map.find_opt k m1) (Map.find_opt k m2)
+    in
+    Map.for_all (fun k _ -> correct_for_key k) m1
+    && Map.for_all (fun k _ -> correct_for_key k) m2
+
+  let union_valid f m1 m2 = Map.valid (Map.union f m1 m2)
+
+  let union f m1 m2 =
+    let mu = Map.union f m1 m2 in
+    m1
+    |> Map.for_all (fun k v1 ->
+           match Map.find_opt k m2 with
+           | None -> Map.find_opt k mu =? Some v1
+           | Some v2 -> Map.find_opt k mu =? f k v1 v2)
+    && m2
+       |> Map.for_all (fun k v2 ->
+              match Map.find_opt k m1 with
+              | None -> Map.find_opt k mu =? Some v2
+              | Some _ -> (* already checked *) true)
+    && mu |> Map.for_all (fun k _ -> Map.mem k m1 || Map.mem k m2)
+
+  let compare_vs_equal m1 m2 =
+    Map.compare V.compare m1 m2 = 0 <=> Map.equal V.equal m1 m2
+
+  let compare_refl = reflexive (Map.compare V.compare)
+
+  let compare_antisym = antisymmetric (Map.compare V.compare)
+
+  let compare_trans = transitive (Map.compare V.compare)
+
+  let equal_vs_bindings m1 m2 =
+    Map.equal V.equal m1 m2
+    <=> List.equal equal_bindings (m1 |> Map.bindings) (m2 |> Map.bindings)
+
+  let for_all_vs_bindings f m =
+    Map.for_all f m <=> List.for_all (fun (k, v) -> f k v) (m |> Map.bindings)
+
+  let exists_vs_bindings f m =
+    Map.exists f m <=> List.exists (fun (k, v) -> f k v) (m |> Map.bindings)
+
+  let iter_vs_fold m =
+    (* See comment on [Set_specs.iter_vs_fold] *)
+    let int_of_value (v : Value.t) = match v with A -> 1 | B -> 2 in
+    let result_by_iter =
+      let state = ref 0 in
+      Map.iter (fun k v -> state := k + (v |> int_of_value) + !state) m;
+      !state
+    in
+    let result_by_fold =
+      Map.fold (fun k v state -> k + (v |> int_of_value) + state) m 0
+    in
+    result_by_iter = result_by_fold
+
+  let filter_valid f m = Map.valid (Map.filter f m)
+
+  let filter f m =
+    let mf = Map.filter f m in
+    Set.subset (Map.keys mf) (Map.keys m)
+    && m
+       |> Map.for_all (fun k v ->
+              Map.find_opt k mf =? if f k v then Some v else None)
+
+  let filter_map_valid f m = Map.valid (Map.filter_map f m)
+
+  let filter_map f m =
+    let mf = Map.filter_map f m in
+    Set.subset (Map.keys mf) (Map.keys m)
+    && m |> Map.for_all (fun k v -> Map.find_opt k mf =? f k v)
```

Comment:
I might have found this clearer if the two sides of the `=?` were swapped, but feel free to disregard this